lib/WUI/nhttp/file_info.cpp

```diff
@@ -37,153 +143,83 @@ FileInfo::FileInfo(const char *filename, bool can_keep_alive, bool after_upload)
-                    filename_escaped, name_escaped);
-                return written;
-            });
-            return Step { 0, written, Continue() };
-        }
-        // Run out of files, end the JSON and finish up.
-        size_t written = render_chunk(handling, output, output_size - last_chunk_len, [&](uint8_t *output, size_t output_size) {
-            return snprintf(reinterpret_cast<char *>(output), output_size, "]}");
-        });
-        if (can_keep_alive) {
-            memcpy(output + written, LAST_CHUNK, last_chunk_len);
-            written += last_chunk_len;
+    size_t written = 0;
+    const ConnectionHandling handling = can_keep_alive ? ConnectionHandling::ChunkedKeep : ConnectionHandling::Close;
+
+    if (holds_alternative<Uninitialized>(renderer)) {
+        // We have not been initialized yet. Try deciding if we are listing a
+        // directory or a file (or if there's an error).
+        //
+        // At this point we not only create the right renderer, we also produce
+        // the right headers.
+
+        struct stat finfo;
+
+        if (DIR *dir_attempt = opendir(filename); dir_attempt) {
+            renderer = DirRenderer(this, dir_attempt);
+            written = write_headers(output, output_size, after_upload ? Status::Created : Status::Ok, ContentType::ApplicationJson, handling);
+        } else if (stat(filename, &finfo) == 0) {
+            renderer = FileRenderer(this, finfo.st_size);
+            written = write_headers(output, output_size, after_upload ? Status::Created : Status::Ok, ContentType::ApplicationJson, handling);
+        } else if (strcmp(filename, "/usb") == 0) {
+            // We are trying to list files in the root and it's not there -> USB is missing.
+            // Special case it, we return empty list of files.
+            renderer = DirRenderer(); // Produces empty file list
+            written = write_headers(output, output_size, after_upload ? Status::Created : Status::Ok, ContentType::ApplicationJson, handling);
+        } else {
+            return Step { 0, 0, StatusPage(Status::NotFound, can_keep_alive) };
         }
-        dir.reset();
-        return Step { 0, written, Terminating::for_handling(handling) };
     }
+    // Note: The above falls through and tries to put at least part of the answer into the rest of the current packet with headers.
 
-    /*
-     * Not having a directory listing running _yet_. So we try that. If it's
-     * not a directory, this'll fail and we will try a file.
-     */
-    if (DIR *dir_attempt = opendir(filename); dir_attempt) {
-        dir.reset(dir_attempt);
-
-        size_t written = write_headers(output, output_size, after_upload ? Status::Created : Status::Ok, ContentType::ApplicationJson, handling);
-
+    // Process both the renderers in uniform way.
+    JsonRenderer *rend = get_if<DirRenderer>(&renderer);
+    if (rend == nullptr) {
+        rend = get_if<FileRenderer>(&renderer);
```

Comment:
```suggestion
    rend = rend ?: get_if<FileRenderer>(&renderer);
```

Reply:
I'm not entirely sure if this is more readable, but :shrug:

Reply:
The proposed `?:` is definitely LESS readable than the original `if`. Moreover, the ternary operator was designed to return one of two options. This is not the case - we only want `rend` to be set to something when it comes with a `nullptr`, otherwise don't touch it.